compiler/passes/codegen.cpp

```diff
@@ -806,7 +806,128 @@ static void protectNameFromC(Symbol* sym) {
 
 
 // TODO: Split this into a number of smaller routines.<hilde>
-static void codegen_header(bool isHeader) {
+static void codegen_defn() {
+  GenInfo* info = gGenInfo;
+  Vec<const char*> cnames;
+  Vec<TypeSymbol*> types;
+  std::vector<FnSymbol*> functions;
+  Vec<VarSymbol*> globals;
+
+  // reserved symbol names that require renaming to compile
+#include "reservedSymbolNames.h"
+
+  //
+  // collect types and apply canonical sort
+  //
+  forv_Vec(TypeSymbol, ts, gTypeSymbols) {
+    if (ts->defPoint->parentExpr != rootModule->block) {
+      types.add(ts);
+    }
+  }
+  qsort(types.v, types.n, sizeof(types.v[0]), compareSymbol);
+
+  //
+  // collect globals and apply canonical sort
+  //
+  forv_Vec(VarSymbol, var, gVarSymbols) {
+    if (var->defPoint->parentExpr != rootModule->block &&
+        toModuleSymbol(var->defPoint->parentSymbol)) {
+      globals.add(var);
+    }
+  }
+  qsort(globals.v, globals.n, sizeof(globals.v[0]), compareSymbol);
+
+  //
+  // collect functions and apply canonical sort
+  //
+  forv_Vec(FnSymbol, fn, gFnSymbols) {
+    functions.push_back(fn);
+  }
+  std::sort(functions.begin(), functions.end(), compareSymbol2);
```

Comment:
There's some duplication here with codegen_header().  Would it be possible to send in pointers to the Vec<>s you're creating to both functions, so that whichever is called second can benefit from the sorting and adding into the vector that is done in the one that is called first?